tidy3d/plugins/invdes/transformation.py

```diff
@@ -0,0 +1,73 @@
+# transformations applied to design region
+
+import typing
+import abc
+
+import jax.numpy as jnp
+import pydantic.v1 as pd
+
+import tidy3d as td
+import tidy3d.plugins.adjoint.utils.filter as tda_filter
+
+from .base import InvdesBaseModel
+
+
+class AbstractTransformation(InvdesBaseModel, abc.ABC):
+    """Base class for transformations."""
+
+
+class FilterProject(InvdesBaseModel):
+    """Transformation involving convolution by a conic filter followed by a ``tanh`` projection.
+
+    .. image:: ../../_static/img/filter_project.png
+
+    """
+
+    radius: float = pd.Field(
```

Comment:
Wouldn't it be more intuitive to have the user provide the physical dimensions and have the parametrization take care of the `sqrt(3)` factor internally?

Reply:
Yea in retrospect, this might have been better. I think we discussed it at some point internally.

This transformation is strongly related to the built-in [adjoint one](https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/plugins/adjoint/utils/filter.py#L22-L33). So some of the reasoning was to not create two projection functions with different conventions. We can't really change that one due to backwards compatibility, so I wanted to just make them consistent.

That being said, it turns out (in my observation) that the feature sizes are much more influenced by the `length_scale` of the dilation-erosion penalty. The radius of the filter is almost more to "help" move the optimization towards larger feature sizes, but the value itself doesn't have much of an effect on things. Would be interesting to get a better perspective on this though if you have any thoughts.



So one line of

Reply:
> So one line of

Guess something is missing here? 😄 

> The radius of the filter is almost more to "help" move the optimization towards larger feature sizes, but the value itself doesn't have much of an effect on things. Would be interesting to get a better perspective on this though if you have any thoughts.

Yeah that's true, although it also depends on the specific filter. Gaussian and cone definitely, but a uniform kernel is somewhat stricter (although the factor doesn't apply there so maybe that's a moot point). In general one can also apply erosion/dilation and derived morphological operations directly in the parametrization (not as a penalty), which can also lead to strict feature sizes. It's a bit tricky to decide which method to pick as a default, as it always depends on the user's requirements, so perhaps it is best to stick to what's found most commonly in literature for now, as that will be what people are familiar with.

Regarding the `radius` argument though, perhaps it would be best to be explicit about this in the documentation, that it doesn't really enforce feature sizes? As you said, it does to some extent, but in my opinion it really is just a regularization technique for the optimization (although people don't usually frame it that way).

Reply:
> In general one can also apply erosion/dilation and derived morphological operations directly in the parametrization (not as a penalty), which can also lead to strict feature sizes.

Yes, we use the filter for our erosion and dilation operations actually.

>  perhaps it is best to stick to what's found most commonly in literature for now, as that will be what people are familiar with.

The literature (at least O Sigmund papers) tend to use a radius R for the "filter" radius I believe. So that's the same as the `radius` field. At least last I remember.

> Regarding the radius argument though, perhaps it would be best to be explicit about this in the documentation, that it doesn't really enforce feature sizes? As you said, it does to some extent, but in my opinion it really is just a regularization technique for the optimization (although people don't usually frame it that way).

Agree, probably we can be more descriptive here. I'll add that.